go/vt/vtgate/evalengine/fn_string.go

```diff
@@ -0,0 +1,284 @@
+	arg, err := call.arg1(env)
+	if err != nil {
+		return nil, err
+	}
+	switch e := arg.(type) {
+	case nil:
+		return nil, nil
+	case *evalBytes:
+		if sqltypes.IsBinary(e.sqlType()) {
+			return newEvalInt64(int64(len(e.bytes))), nil
+		}
+		coll := collations.Local().LookupByID(e.col.Collation)
+		count := collations.Length(coll, e.bytes)
+		return newEvalInt64(int64(count)), nil
+	default:
+		return newEvalInt64(int64(len(e.toRawBytes()))), nil
+	}
+}
+
+func (call *builtinCharLength) typeof(env *ExpressionEnv) (sqltypes.Type, typeFlag) {
+	_, f := call.Arguments[0].typeof(env)
+	return sqltypes.Int64, f
+}
+
+func (call *builtinLength) eval(env *ExpressionEnv) (eval, error) {
+	arg, err := call.arg1(env)
+	if err != nil {
+		return nil, err
+	}
+	if arg == nil {
+		return nil, nil
+	}
+	return newEvalInt64(int64(len(arg.toRawBytes()))), nil
+}
+
+func (call *builtinLength) typeof(env *ExpressionEnv) (sqltypes.Type, typeFlag) {
+	_, f := call.Arguments[0].typeof(env)
+	return sqltypes.Int64, f
+}
+
+func (call *builtinBitLength) eval(env *ExpressionEnv) (eval, error) {
+	arg, err := call.arg1(env)
+	if err != nil {
+		return nil, err
+	}
+	if arg == nil {
+		return nil, nil
+	}
+	return newEvalInt64(int64(len(arg.toRawBytes())) * 8), nil
+}
+
+func (call *builtinBitLength) typeof(env *ExpressionEnv) (sqltypes.Type, typeFlag) {
+	_, f := call.Arguments[0].typeof(env)
+	return sqltypes.Int64, f
+}
+
+func (call *builtinASCII) eval(env *ExpressionEnv) (eval, error) {
+	arg, err := call.arg1(env)
+	if err != nil {
+		return nil, err
+	}
+	if arg == nil {
+		return nil, nil
+	}
+
+	b, ok := arg.(*evalBytes)
+	if !ok {
+		b = evalToBinary(arg)
+	}
+	if len(b.bytes) == 0 {
+		return newEvalInt64(0), nil
+	}
+	return newEvalInt64(int64(b.bytes[0])), nil
+}
+
+func (call *builtinASCII) typeof(env *ExpressionEnv) (sqltypes.Type, typeFlag) {
+	_, f := call.Arguments[0].typeof(env)
+	return sqltypes.Int64, f
+}
+
+type builtinRepeat struct {
+	CallExpr
+}
+
+func (call *builtinRepeat) eval(env *ExpressionEnv) (eval, error) {
+	arg1, arg2, err := call.arg2(env)
+	if err != nil {
+		return nil, err
+	}
+	if arg1 == nil || arg2 == nil {
+		return nil, nil
+	}
+
+	text, ok := arg1.(*evalBytes)
+	if !ok {
+		text, err = evalToText(arg1, env.DefaultCollation, true)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	repeat := evalToNumeric(arg2).toInt64().i
+	if repeat < 0 {
+		repeat = 0
+	}
+	return newEvalText(bytes.Repeat(text.bytes, int(repeat)), text.col), nil
+}
+
+func (call *builtinRepeat) typeof(env *ExpressionEnv) (sqltypes.Type, typeFlag) {
+	_, f1 := call.Arguments[0].typeof(env)
+	// typecheck the right-hand argument but ignore its flags
+	call.Arguments[1].typeof(env)
+	return sqltypes.VarChar, f1
+}
+
+func (c *builtinCollation) eval(env *ExpressionEnv) (eval, error) {
+	arg, err := c.arg1(env)
+	if err != nil {
+		return nil, err
+	}
+
+	col := collations.Local().LookupByID(evalCollation(arg).Collation)
```

Comment:
can `col` ever be `nil` if the given collation does not exist?

Reply:
No, the collations are always tested before translation. There are no missing collations in the engine.

Reply:
Got it!